tripal_chado/src/Plugin/ChadoBuddy/ChadoCvtermBuddy.php

```diff
@@ -0,0 +1,1016 @@
+   *   a cvterm with a feature the basetable=feature and cvterm_id is added to the
+   *   feature_cvterm table.
+   * @param int $record_id
+   *   The primary key of the basetable to associate the cvterm with.
+   * @param ChadoBuddyRecord $cvterm
+   *   A cvterm object returned by any of the *Cvterm() in this service.
+   * @param $options
+   *   'pkey': Looking up the primary key for the base table is costly. If it is
+   *           known, then pass it in as this option for better performance.
+   *   Also pass in any other columns used in the linking table, some of which may
+   *   have a NOT NULL constraint. See the table below for a list of which of
+   *   the following may be required: 'pub_id', 'is_not', 'rank', 'cvterm_type_id'.
+   *
+   *   Chado 1.3 defines these columns in the various linking tables:
+   *   ^ table                       ^ pub_id   ^ is_not      ^ rank        ^ cvterm_type_id ^
+   *   | analysis_cvterm             | -absent  | has default | has default | -absent        |
+   *   | cell_line_cvterm            | not null | -absent     | has default | -absent        |
+   *   | environment_cvterm          | -absent  | -absent     | -absent     | -absent        |
+   *   | expression_cvterm           | -absent  | -absent     | has default | not null       |
+   *   | feature_cvterm              | not null | has default | has default | -absent        |
+   *   | library_cvterm              | not null | -absent     | -absent     | -absent        |
+   *   | organism_cvterm             | not null | -absent     | has default | -absent        |
+   *   | phenotype_comparison_cvterm | not null | -absent     | has default | -absent        |
+   *   | phenotype_cvterm            | -absent  | -absent     | has default | -absent        |
+   *   | stock_cvterm                | not null | has default | has default | -absent        |
+   *   | stock_relationship_cvterm   | yes null | -absent     | -absent     | -absent        |
+   *
+   * @return bool
+   *   Returns TRUE if successful.
+   *   Both the cvterm and the chado record indicated by $record_id MUST ALREADY EXIST.
+   *
+   * @throws Drupal\tripal_chado\ChadoBuddy\Exceptions\ChadoBuddyException
+   *   If an error is encountered.
+   */
+  public function associateCvterm(string $base_table, int $record_id, ChadoBuddyRecord $cvterm, array $options = []) {
+    $linking_table = $base_table . '_cvterm';
+
+    // Get the primary key of the base table
+    $base_pkey_col = $options['pkey'] ?? NULL;
+    if (!$base_pkey_col) {
+      $schema = $this->connection->schema();
+      $base_table_def = $schema->getTableDef($base_table, ['format' => 'Drupal']);
+      $base_pkey_col = $base_table_def['primary key'];
+    }
+    $fields = [
+      'cvterm_id' => $cvterm->getValue('cvterm.cvterm_id'),
+      $base_pkey_col => $record_id,
+    ];
+    // Add in any of the other columns for the linking table.
```

Comment:
What about if the user chooses not to care about performance and doesn't add the keys needed for their particular linking table? Shouldn't we then look them up?

What about having an option `lookup_columns` that is TRUE by default unless other keys in addition to the pkey are supplied? So the default would be determined as follows:
```php
$lookup_columns = $options['lookup_columns'] ?? TRUE;
$possible_columns = ['pub_id','rank','is_not','cvterm_type_id'];
$columns_defined = in_array($possible_columns, $options);
if ($lookup_columns === TRUE && count($columns_defined) > 0) {
  $lookup_columns = FALSE;
}
```

Reply:
Resolving as I have added this.